book/art/fullnode.bob

```diff
@@ -1,27 +1,30 @@
-             .---------------------------.
-             |  Fullnode                 |
-             |                           |
- .--------.  |  .------------------.     |
- |        |---->|                  |     |
- | Client |  |  | JSON RPC Service |     |
- |        |<----|                  |     |
- `----+---`  |  `------------------`     |
-      |      |     ^                     |    .------------------.
-      |      |     |  .----------------. |    | Validators       |
-      |      |     |  | Gossip Service +----->|                  |
-      |      |     |  `--------+-------` |    |  .------------.  |
-      |      |     |         ^ |         |    |  |            |  |
-      |      |     |         | v         |    |  | Upstream   |  |
-      |      |  .--+---.   .-+---.       |    |  | Validators |  |
-      |      |  | Bank |<--| TVU |<--------------+            |  |
-      |      |  `------`   `-----`       |    |  `------------`  |
-      |      |     ^                     |    |                  |
-      |      |     |                     |    |  .------------.  |
-      |      |  .--+--.   .-----------.  |    |  |            |  |
-      `-------->| TPU +-->| Broadcast +--------->| Downstream |  |
-             |  `-----`   | Service   |  |    |  | Validators |  |
-             |            `-----------`  |    |  |            |  |
-             |                           |    |  `------------`  |
-             `---------------------------`    |                  |
-                                              `------------------`
-
+             .---------------------------------.
+             |  Fullnode                       |
+             |                                 |
+ .--------.  |  .------------------.           |
+ |        |---->|                  |           |
+ | Client |  |  | JSON RPC Service |           |
+ |        |<----|                  |           |
+ `----+---`  |  `------------------`           |
+      |      |     ^    .----------------.     |
+      |      |     |    | Gossip Service |     |  .------------------.
+      |      |     |    | (syncs votes & |<------>| Validators       |
+      |      |     |    |  contact info) |     |  |                  |
+      |      |     |    `----------------`     |  |                  |
+      |      |     |                           |  |  .------------.  |
+      |      |  .--+---.  .------.  .--------. |  |  |            |  |
+      |      |  | Bank |<-|Replay|  |Network |<------+ Upstream   |  |
```

Comment:
What is "Network Observer" and why is it in Fullnode?

Reply:
Network observer is currently the part of TVU (fetch_blob, recv_window, retransmit, write to ledger etc). Everything in TVU minus replay stage..

Reply:
Yikes, we'll have to rename that. But if consistent with the implementation, good to see it here.

Reply:
It's not implemented yet this way. I think, there'll be a refactor after bank forking is merged. @rob-solana , do you think we'll use this name in the implementation?

Reply:
The "observer" would be some entity outside the cluster, right? not a component within a fullnode within a cluster.  How about, Block Explorer Service?  Maybe we give it a more general name if it is ever used for anything other than the block explorer?

Reply:
May be it can use a better name. This component is part of the fullnode (fetch_blob, recv_window, retransmit, write to ledger etc).

Reply:
Sounds good.

Reply:
@CriesofCarrots called it Entry Stream Stage here https://github.com/solana-labs/solana/pull/2746. Can you two sync and make sure both PRs use the same name?

Reply:
Correct me if I'm wrong, @pgarg66, but I don't think "Network Observer" is intended to refer to EntryStream or Block Explorer Service. If I understand the diagram correctly, this is the start of the Tvu pipeline -- blob fetch and retransmit.

Reply:
@CriesofCarrots that's correct. It's not specific to Block Explorer. It's a split of current TVU (TVU minus replay stage).

The idea is that ReplayStage will not be fed entries from a channel (it used to be from retransmit stage). It'll get notifications whenever the ledger is updated. It'll read the ledger, process the entries in bank and vote on the fork of choice.

The ledger will be updated by the "network observer". It's blob fetcher will receive blobs from peers/leader, run it through retransmit stage (window service, repairs, and retransmit), and write to ledger.

The leader will also write to the ledger from its broadcast service.

Reply:
This diagram reflects what's currently implemented. Some of the changes here are indeed implemented, but what you're talking about here isn't. Is what you've described documented in any architectural proposal?

Reply:
You are right. This is not in the current code base. This came up yesterday in the bank fork discussions. I can submit it as a architectural proposal. Do we have guidelines on that process?

Reply:
https://github.com/solana-labs/solana/blob/master/CONTRIBUTING.md#proposing-architectural-changes

Reply:
👍

Reply:
For renaming "Network Observer", how is "Peer Transaction Receiver"?